text/0000-project-toolchain-shims.md

```diff
@@ -0,0 +1,69 @@
+- Feature Name: project_toolchain_shims
+- Start Date: 2018-09-19
+- RFC PR: (leave this empty)
+- Notion Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Automatic management of project toolchain shims will be handled by the `yarn` and `npm` shims, which will scan the `package.json` files of the project and its immediate dependencies to determine which shims to create.
+
+# Motivation
+[motivation]: #motivation
+
+Automatic management of shims is a core competency of Notion, in that its abstraction of the Node ecosystem workflow is incomplete without it.  Automatic management of user toolchain shims is provided by the `notion install` command, but no process for automatically managing project toolchain shims has yet been described.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+If this feature works as intended, there is little to teach.  The only significant divergence from typical JS workflows is that `./node_modules/.bin/` can be omitted when calling tools in the project toolchain:
+
+```
+$ ./node_modules/.bin/mocha  # The non-Notion way
+$ mocha  # The Notion way
+```
+
+# Details
+[details]: #details
+
+Just prior to a successful completion inside a project, the `yarn` and `npm` shims will "scan" for dependency binaries and attempt to create shims for them.  A "successful" completion, in this sense, means that the underlying tool was successfully run; it does *not* require that the tool returned a status of 0.  This covers all potential scenarios in which either of those tools create new dependency binaries.
+
+This scan will be a two-level process.  First, the project's `package.json` will be inspected for all dependencies and devDependencies.  Second, each of those dependencies' `package.json` files will be inspected for binaries to shim.  For each binary, Notion will attempt to create a shim symlink, catching `EEXISTS` and considering it a success.  Any other error will result in an appropriate message to `stderr`, but not halt the scanning/shimming process.
+
+This same scan will also occur when pinning a version of Node using `notion use`, to ensure that shims are created for packages installed prior to setting up a toolchain for the project.  Finally, the scan will be available via a new `notion autoshim .` command.
```

Comment:
I'd cut the `autoshim` command for now. I think we should see how the tool feels with shim management kept entirely under the hood. We can always consider adding shim management commands back in after we've gotten some experience using Notion for a while.

It's not _impossible_ to create or delete a shim by hand without a subcommand, it's just manual and icky enough to dissuade users from doing it casually. And you can always trigger autoshimming by running `yarn` or `npm i`.

But I'm now coming to the opinion that we should start by eliminating the `notion shim` command completely, and drop the `autoshim` command from this RFC.

What do you think?

Reply:
I agree that `notion shim` isn't a user-facing command, but (at least for now) I think there is some value in it for Notion developers.  For example, I found myself deleting third-party shims quite frequently as part of implementing auto-shimming.
Between that and the fact that `notion shim auto` is already implemented and working, I'd prefer to punt this for now.  It's easy enough to remove the command later, and in the mean time, we may find value in it for scripting or some such.  (I liken `notion shim` to git's "plumbing" commands, to a degree.)

Reply:
@benblank I don't want to start shipping things to users just because they're already implemented -- the point of RFCs is to try to get to consensus before we ship things. And we need to be careful about shipping user-visible features with the expectation of removign them -- removing features is very costly once you have users, and we're getting close to launching and encouraging adoption.

At the same time, I hear what you're saying about the usefulness of internal features for developers.

Let's use Cargo's feature flags functionality to put features that are only for developers behind a flag. And let's put the entire `notion shim` command behind a feature flag; I'm not convinced it's an appropriate feature for end users. If we decide later that it is, it'll already be in the codebase and easy to enable.

Reply:
BTW your point about "plumbing" is good too. Probably the best way to think about plumbing commands is that they're for external automation (people writing extensions or scripts).

We definitely should expose plumbing commands but they're lower-level and hard to change, so we can be a little more cautious about exposing them. Starting with a feature flag seems like a good first step, and then we should think about the UI of plumbing commands. (For example, maybe we want to mark them explicitly as such in the UI, e.g. `notion plumbing shim ...`? But we don't need to decide that here for this RFC.)

Reply:
Putting `notion shim` behind a feature flag sounds like a great idea.  I'll update the implementation PR as soon as I figure out how to do that.  ;-)

Reply:
@benblank https://doc.rust-lang.org/cargo/reference/manifest.html#the-features-section is a good place to start.